fl4health/preprocessing/autoencoders/dim_reduction.py

```diff
@@ -26,59 +28,113 @@ def __repr__(self) -> str:
-
-    def __init__(
-        self,
-        checkpointing_path: Path,
-    ) -> None:
-        super().__init__(checkpointing_path)
+class AeProcessor(AutoEncoderProcessing):
+    """
+    Transformer processor to encode the data using basic autoencoder.
+    """
 
     def __call__(self, sample: torch.Tensor) -> torch.Tensor:
         # This transformer is called for the input samples after they are transfered into torch tensors.
         embedding_vector = self.autoencoder.encode(sample.to(self.device))
         return embedding_vector.clone().detach()
 
 
-class VaeProcessor(Processing):
-    """Transformer processor to encode the data using VAE encoder."""
-
+class VaeProcessor(AutoEncoderProcessing):
     def __init__(
         self,
         checkpointing_path: Path,
-        include_variance: bool = False,
+        return_mu: bool = False,
     ) -> None:
+        """
+        Transformer processor to encode the data using VAE encoder.
+
+        Args:
+            checkpointing_path (Path): Path to the saved model.
+            return_mu (bool, optional): If true, only mu is returned. Defaults to False.
+        """
         super().__init__(checkpointing_path)
-        self.include_variance = include_variance
+        self.return_mu = return_mu
 
     def __call__(self, sample: torch.Tensor) -> torch.Tensor:
         # This transformer is called for the input samples after they are transfered into torch tensors.
         mu, logvar = self.autoencoder.encode(sample.to(self.device))
-        if self.include_variance:
-            return torch.cat(mu.clone().detach(), logvar.clone().detach())
-        return mu.clone().detach()
+        if self.return_mu:
+            return mu.clone().detach()
+        # By default returns cat(mu,logvar).
+        # Contatation is performed on the last dimention which for both the batched data and single data
+        # is the latent space dimention.
+        return torch.cat((mu.clone().detach(), logvar.clone().detach()), dim=-1)
 
 
-class CvaeFixedConditionProcessor(Processing):
+class CvaeFixedConditionProcessor(AutoEncoderProcessing):
     """Transformer processor to encode the data using CVAE encoder with client-specific condition."""
 
     def __init__(
         self,
         checkpointing_path: Path,
         condition: torch.Tensor,
-        include_variance: bool = False,
+        return_mu: bool = False,
+        batch_size: int = 1,
     ) -> None:
+        super().__init__(checkpointing_path)
         self.condition = condition
-        self.include_variance = include_variance
+        self.return_mu = return_mu
         assert (
             self.condition.dim() == 1
         ), f"Error: condition should be a 1D vector instead of a {self.condition.dim()}D tensor."
-        super().__init__(checkpointing_path)
+        # Assuming data is "batch first"
+        # If we are processing a batch of data, condition should be repeated for each sample.
+        if batch_size > 1:
```

Comment:
I think this is a good idea, but I think we want to do this "dynamically" in the `__call__` function below because the batch size may not be constant. 

A simple example of this is if the dataset is perfectly divisible by the batch size, your last batch may not have the same size as the previous ones (unless you set drop_last = True in your dataloader). However, there are other settings where your batch size fluctuates, such as doing Poisson sampling to form your batches.

So, I would do something like `self.condition = condition.to(self.device)` in the `__init__`, remove `batch_size` as a property, and then do something like 
```
# assuming batch is the first dimension
sample_batch_size = sample.shape[0]
batch_condition = self.condition.repeat(sample_batch_size, 1)
mu, logvar = self.autoencoder.encode(sample.to(self.device), batch_condition)
```
in the `__call__` function

Reply:
Right, that is a very good point! Reshaping the condition in init seemed more efficient but won't cover these situations.